### DISCUSSION TITLE
MQTT using mbed-ce

### DISCUSSION THREAD

**Discussion by timIdeaTech** [mbed-ce/mbed-os#269] [Q&A] [answered]
Looking to see if anybody has implemented MQTT on mbed-ce?

**Comment by JohnK1987** [comment, marked as answer]
I think @JojoS62 
https://github.com/JojoS62/mbed-ce-test-apps/tree/test-JZ_F407VE/mbed-libs

**Comment by timIdeaTech** [comment]
Thanks! A few useful things in there.

I've been adapting the ARMmbed/mbed-mqtt library and got it compiling this morning to the point of a -3004 connect error.

**Reply by JojoS62** [reply]
This seams to be a network connect problem.
I have also an app for the MQTT code. 

https://github.com/JojoS62/mbed-ce-test-apps/tree/test-JZ_F407VE/app_Ethernet

**Reply by timIdeaTech** [reply]
@JojoS62 ,  I recently had some time to try out your MQTT library and it seems clean and easy to use. Is this a work in progress? 

I tried setting QoS1 & 2 for PubMessages. QoS1 doesn't seem to actually set the bit and QoS2 doesn't seem to publish the message, or at least it doesn't get through. Also, shouldn't there be a "retain" option in the PubMessage struct?

I haven't had a chance to do much digging yet but did notice a few "TODO:" comments, hence my question above. No judgement here, JohnK1987 and you threw the link out there with no caveats ;^)

I'm currently using MQTT Explorer and mosquito on Windows 11 for reference.

**Comment by timIdeaTech** [comment]
Thanks again.

It was just an issue with a gethostbyname() call not finding our local subnet names. I seem to have my modified version of the ARMmbed/mbed-mqtt library working with mbed-ce. Learning more about CMake as I go. I will also need MQTT-SN down the road but I will give you library a go. 

Is your mbed-mqtt library a condensed version of the ARMmbed library?
